crawler.py

```diff
@@ -97,25 +109,53 @@ def run(self):
 
 		logging.info("Start the crawling process")
 
-		while len(self.tocrawl) != 0:
-			self.__crawling()
+		if self.num_workers == 1:
+			while len(self.urls_to_crawl) != 0:
+				current_url = self.urls_to_crawl.pop()
+				self.crawled_or_crawling.add(current_url)
+				self.__crawl(current_url)
+		else:
+			event_loop = asyncio.get_event_loop()
+			try:
+				while len(self.urls_to_crawl) != 0:
+					executor = concurrent.futures.ThreadPoolExecutor(max_workers=self.num_workers)
+					event_loop.run_until_complete(self.crawl_all_pending_urls(executor))
+			finally:
+				event_loop.close()
```

Comment:
So, here's you'll notice the single-threaded logic is identical (although I did lift 2 lines out of the `self.__crawl` method).